### PR TITLE
feat: mysql-k8s integration

### PR DESCRIPTION
**Pull request by i-chvets**
Summary of changes:
- Initial implementation and integration test.
- Mysql-k8s relation is not working in test: data is not sent and events are not triggred.
- Unit test are passing.
- Integration test are not passing.
- Making it a draft PR.

### CODE REVIEW

File: charms/katib-db-manager/requirements.in [hunk 1]
```diff
@@ -1,5 +1,5 @@
-ops
+ops==2.2.0
 oci-image
 charmed-kubeflow-chisme
 lightkube
-lightkube-models
\ No newline at end of file
+lightkube-models==1.25.4.4
```
Comment (beliaev-maksim) [inline charms/katib-db-manager/requirements.in:5]:
```suggestion
lightkube-models==1.25.4.4

```

File: charms/katib-db-manager/src/charm.py [hunk 2]
```diff
@@ -136,17 +158,132 @@ def _check_leader(self):
...
+        # relying on KeyError to ensure that relational-db relation is not present
+        try:
+            relation = self.model.get_relation("relational-db")
+            if relation:
+                self.logger.warnning(
+                    "Failed to create mysql relation due to existing relational-db relation."
+                )
+                return
+        except KeyError:
+            # relational-db relation was not found, proceed
```
Comment (beliaev-maksim) [inline charms/katib-db-manager/src/charm.py:173]: no, we need to add warning and juju status telling that mysql is deprecated

File: charms/katib-db-manager/src/charm.py [hunk 3]
```diff
@@ -136,17 +158,132 @@ def _check_leader(self):
...
+                f"Missing attribute {err} in mysql relation data: {relation_data}"
+            )
+            # incorrect/incomplete data can be found in mysql relation which can be resolved:
+            # use WaitingStatus
+            raise ErrorWithStatus(
+                f"Incorrect/incomplete data found in relation mysql. See logs", WaitingStatus
+            )
+        except GenericCharmRuntimeError:
+            # at this point it is detected that mysql relation is not established
+            # check for relational-db relation
```
Comment (beliaev-maksim) [inline charms/katib-db-manager/src/charm.py:243]: I think we need two different functions. One will handle mariadb, another mysql. That will help us in future when you will remove deprecated interface. Then we just delete one method. But now it is a spaghetti monster with mixing of all the stuff together :D

File: charms/katib-db-manager/tests/unit/test_operator.py [hunk 4]
```diff
@@ -167,3 +168,104 @@ def test_update_status(
...
+    harness.model.get_relation = MagicMock(
+        side_effect=_get_relation_db_only_side_effect_func
+    )
+    harness.begin()
+    harness.charm.database = database
+    try:
+        harness.charm._get_db_data()
+        assert 0 # expected error, but returned data
+    except ErrorWithStatus as err:
+        assert err.status == WaitingStatus("Waiting for relational-db data")
```
Comment (beliaev-maksim) [inline charms/katib-db-manager/tests/unit/test_operator.py:197]: with pytest raises ?

**Review by beliaev-maksim — CHANGES_REQUESTED** [review]

File: charms/katib-db-manager/requirements.in [hunk 5]
```diff
@@ -1,5 +1,5 @@
-ops
+ops==2.2.0
```
Comment (beliaev-maksim) [inline charms/katib-db-manager/requirements.in:1]: can you please do a bump in a separate PR and ensure all tests are passing

File: charms/katib-db-manager/tests/unit/test_operator.py (hunk 4 above)
Comment (misohu) [reply charms/katib-db-manager/tests/unit/test_operator.py:197]: here an example https://github.com/canonical/mlflow-operator/blob/KF-715-mlflow-v211/tests/unit/test_operator.py#L262

File: charms/katib-db-manager/requirements.in (hunk 5 above)
Comment (beliaev-maksim) [reply charms/katib-db-manager/requirements.in:1]: see https://github.com/canonical/katib-operators/pull/73

File: charms/katib-db-manager/src/charm.py (hunk 2 above)
Comment (i-chvets) [reply charms/katib-db-manager/src/charm.py:173]: Warning will go into log. Status will be overwritten pretty quickly, once we start processing mysql relation data.

File: charms/katib-db-manager/src/charm.py [hunk 6]
```diff
@@ -136,17 +158,132 @@ def _check_leader(self):
...
-    def _check_mysql(self):
+    def _on_mysql_relation(self, event):
+        """Check for existing database relations and process mysql relation if needed."""
+        # check for relational-db relation
+        # relying on KeyError to ensure that relational-db relation is not present
+        try:
+            relation = self.model.get_relation("relational-db")
+            if relation:
+                self.logger.warnning(
+                    "Failed to create mysql relation due to existing relational-db relation."
```
[the remark this answers is not in the document]
Comment (i-chvets) [reply charms/katib-db-manager/src/charm.py:169]: +1

File: charms/katib-db-manager/src/charm.py [hunk 7]
```diff
@@ -136,17 +158,132 @@ def _check_leader(self):
...
+        """Check for existing database relations and process relational-db relation if needed."""
+        # check for mysql relation
+        # relying on KeyError to ensure that mysql relation is not present
+        assert 0
         try:
-            relation = self.model.relations["mysql"][0]
+            relation = self.model.get_relation("mysql")
+            if relation:
+                self.logger.warninig(
+                    "Failed to create relational-db relation due to existing mysql relation"
```
[the remark this answers is not in the document]
Comment (i-chvets) [reply charms/katib-db-manager/src/charm.py:190]: +1

File: charms/katib-db-manager/tests/unit/test_operator.py (hunk 4 above)
Comment (i-chvets) [reply charms/katib-db-manager/tests/unit/test_operator.py:197]: Done.

File: charms/katib-db-manager/tests/unit/test_operator.py [hunk 8]
```diff
@@ -167,3 +168,104 @@ def test_update_status(
...
+    )
+    harness.begin()
+    harness.charm.database = database
+    try:
+        harness.charm._get_db_data()
+        assert 0 # expected error, but returned some data
+    except ErrorWithStatus as err:
+        assert err.status == BlockedStatus(
+            "Incorrect data found in relation relational-db: Missing data. See logs"
+        )
```
[the remark this answers is not in the document]
Comment (i-chvets) [reply charms/katib-db-manager/tests/unit/test_operator.py:219]: Done.

File: charms/katib-db-manager/tests/unit/test_operator.py [hunk 9]
```diff
@@ -167,3 +168,104 @@ def test_update_status(
...
+    )
+    harness.begin()
+    harness.charm.database = database
+    try:
+        harness.charm._get_db_data()
+        assert 0 # expected error, but returned valid data
+    except ErrorWithStatus as err:
+        assert err.status == BlockedStatus(
+            "Incorrect data found in relation relational-db: Missing data. See logs"
+        )
```
[the remark this answers is not in the document]
Comment (i-chvets) [reply charms/katib-db-manager/tests/unit/test_operator.py:241]: Done

File: charms/katib-db-manager/src/charm.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (i-chvets) [reply charms/katib-db-manager/src/charm.py:169]: +1

File: charms/katib-db-manager/src/charm.py [hunk 10]
```diff
@@ -165,123 +168,162 @@ def _on_mysql_relation(self, event):
...
             relation = self.model.get_relation("relational-db")
             if relation:
-                self.logger.warnning(
-                    "Failed to create mysql relation due to existing relational-db relation."
+                self.logger.warning(
+                    "Up-to-date database relation relational-db is already established."
+                )
+                self.logger.warning(f"{MYSQL_WARNING} {UNBLOCK_MESSAGE}")
+                self.model.unit.status = BlockedStatus(
+                    f"{MYSQL_WARNING} {UNBLOCK_MESSAGE} See logs"
```
Comment (beliaev-maksim) [inline charms/katib-db-manager/src/charm.py:176]: 1. I think message is too long for the screen.
2. maybe we should emphasize remove only `mysql` ?

File: charms/katib-db-manager/src/charm.py [hunk 11]
```diff
@@ -165,123 +168,162 @@ def _on_mysql_relation(self, event):
         try:
             relation = self.model.get_relation("relational-db")
             if relation:
-                self.logger.warnning(
-                    "Failed to create mysql relation due to existing relational-db relation."
+                self.logger.warning(
+                    "Up-to-date database relation relational-db is already established."
+                )
+                self.logger.warning(f"{MYSQL_WARNING} {UNBLOCK_MESSAGE}")
+                self.model.unit.status = BlockedStatus(
```
Comment (beliaev-maksim) [inline charms/katib-db-manager/src/charm.py:175]: what is we fire later update status?

how do we get the same BlockedStatus ?

File: charms/katib-db-manager/src/charm.py [hunk 12]
```diff
@@ -165,123 +168,162 @@ def _on_mysql_relation(self, event):
...
             relation = self.model.get_relation("mysql")
             if relation:
-                self.logger.warninig(
-                    "Failed to create relational-db relation due to existing mysql relation"
+                self.logger.warning(
+                    "Failed to create relational-db relation due to existing mysql relation."
+                )
+                self.logger.warning(f"{MYSQL_WARNING} {UNBLOCK_MESSAGE}")
+                self.model.unit.status = BlockedStatus(
+                    f"{MYSQL_WARNING} {UNBLOCK_MESSAGE} See logs"
```
Comment (beliaev-maksim) [inline charms/katib-db-manager/src/charm.py:202]: I really think we need to do something with messaging. 

eg
"mysql and relational-db cannot co-exist. `mysql` is deprecated, please remove it"

in logs we can go extremely explicit.

in status need to play to see the limit. In the worst case "Relations conflict, please see logs"

File: charms/katib-db-manager/src/charm.py [hunk 13]
```diff
@@ -165,123 +168,162 @@ def _on_mysql_relation(self, event):
...
-        assert 0
         try:
             relation = self.model.get_relation("mysql")
             if relation:
-                self.logger.warninig(
-                    "Failed to create relational-db relation due to existing mysql relation"
+                self.logger.warning(
+                    "Failed to create relational-db relation due to existing mysql relation."
+                )
+                self.logger.warning(f"{MYSQL_WARNING} {UNBLOCK_MESSAGE}")
```
Comment (beliaev-maksim) [inline charms/katib-db-manager/src/charm.py:200]: here it is not a warning. Here we go for an error, since we are conflicting

File: charms/katib-db-manager/src/charm.py [hunk 14]
```diff
@@ -8,18 +8,21 @@
...
 from ops.main import main
-from ops.model import ActiveStatus, MaintenanceStatus, ModelError, WaitingStatus
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus, ModelError, WaitingStatus
 from ops.pebble import CheckStatus, Layer
 
 K8S_RESOURCE_FILES = [
     "src/templates/auth_manifests.yaml.j2",
 ]
+MYSQL_WARNING = "Warn: mysql relation is deprecated."
+UNBLOCK_MESSAGE = "Remove mysql or relational-db relation to unblock."
```
Comment (beliaev-maksim) [inline charms/katib-db-manager/src/charm.py:25]: we definitely should not use the same message for both.

I kinda agree that technically that is the solution. But we really want users to disable "mysql"

File: charms/katib-db-manager/src/charm.py [hunk 15]
```diff
@@ -136,17 +161,171 @@ def _check_leader(self):
...
+
+        try:
+            # retrieve relation data
+            relation = self.model.get_relation(relation_name)
+        except KeyError:
+            # relation was not found
+            relation = None
+        if not relation:
+            # relation is not established, raise an error
+            raise GenericCharmRuntimeError(f"Database relation {relation_name} is not established")
```
Comment (beliaev-maksim) [inline charms/katib-db-manager/src/charm.py:229]: or could it be that `self.model.get_relation(relation_name)` returns None ?
```suggestion
            raise GenericCharmRuntimeError(f"Database relation {relation_name} is not established")
```

File: charms/katib-db-manager/src/charm.py [hunk 16]
```diff
@@ -136,17 +161,171 @@ def _check_leader(self):
...
+            # this also validates the expected data by means of KeyError exception
+            db_data["db_name"] = self._database_name
+            db_data["db_username"] = relation_data["user"]
+            db_data["db_password"] = relation_data["root_password"]
+            db_data["katib_db_host"] = relation_data["host"]
+            db_data["katib_db_port"] = relation_data["port"]
+            db_data["katib_db_name"] = relation_data["database"]
+        except (IndexError, StopIteration, KeyError) as err:
+            # failed to retrieve database configuration
+            if len(relation_data) == 0:
```
Comment (beliaev-maksim) [inline charms/katib-db-manager/src/charm.py:253]:
```suggestion
            if not relation_data:
```

File: charms/katib-db-manager/src/charm.py [hunk 17]
```diff
@@ -136,17 +161,171 @@ def _check_leader(self):
...
+            except KeyError as err:
+                self.logger.error(f"Missing attribute {err} in relational-db relation data")
+                # incorrect/incomplete data can be found in mysql relation which can be
+                # resolved: use WaitingStatus
+                raise ErrorWithStatus(
+                    "Incorrect/incomplete data found in relation relational-db. See logs",
+                    WaitingStatus,
+                )
+        # report if there was no data populated
+        if len(db_data) == 0:
```
Comment (beliaev-maksim) [inline charms/katib-db-manager/src/charm.py:298]:
```suggestion
        if not db_data:
```

File: charms/katib-db-manager/src/charm.py (hunk 11 above)
Comment (beliaev-maksim) [reply charms/katib-db-manager/src/charm.py:175]: is it line 380?
what if other event occurs? Will we enter this method as well?
```
    def _on_update_status(self, event):
        """Update status actions."""
        # skip update status processing in case of BlockedStatus
        if self.model.unit.status == BlockedStatus:
            return
```

File: charms/katib-db-manager/src/charm.py (hunk 15 above)
Comment (i-chvets) [reply charms/katib-db-manager/src/charm.py:229]: Actually I missed it here, it should also have `or empty`.
It was too complex to cover all error cases, so I think message indicating `not established or empty` should suffice.

File: charms/katib-db-manager/src/charm.py (hunk 11 above)
Comment (i-chvets) [reply charms/katib-db-manager/src/charm.py:175]: Re-writing it according ot Pietro's suggestion:
> you'd need an `isinstance(self.model.unit.status, BlockedStatus)`, but yeah.
That would mean you'd never recover from a blockedstatus though

[3h ago](https://chat.canonical.com/canonical/pl/u5gwsjesapf5ujzmdt9y5g7ano)
unless you set status from somewhere else as well, that is

File: charms/katib-db-manager/src/charm.py (hunk 10 above)
Comment (i-chvets) [reply charms/katib-db-manager/src/charm.py:176]: Looks like it fits. We also duplicate it in logs.

Comment (i-chvets) [reply charms/katib-db-manager/src/charm.py:176]: I got it shorter:
```
MYSQL_WARNING = "Relation mysql is deprecated."
UNBLOCK_MESSAGE = "Remove deprecated mysql relation to unblock."
. . .
self.logger.error(f"{MYSQL_WARNING} {UNBLOCK_MESSAGE}")
self.model.unit.status = BlockedStatus(f"{UNBLOCK_MESSAGE} See logs")
```

**Review by beliaev-maksim — APPROVED** [review]
